TBB device

new2f7 opened the pull request:
I hereby propose a new device which uses the **Intel Threading Building Blocks** library for scheduling (as suggested [here](https://github.com/pocl/pocl/issues/330#issuecomment-481997802)). The intention of this device is to provide an alternative to the pthread device, which is:
- potentially faster (Intel OpenCL uses TBB)
- more configurable, such that workgroup scheduling best matches the application's needs

Similar to [OpenMP's `schedule` clause](https://www.openmp.org/spec-html/5.0/openmpsu41.html#x64-1290002.9.2), a [partitioner](https://www.threadingbuildingblocks.org/docs/help/reference/algorithms/partitioners.html) and a [grain size](https://www.threadingbuildingblocks.org/docs/help/tbb_userguide/Controlling_Chunking.html) can be specified via environment variables. See below for background information and benchmark results.

**I would be grateful for a code review and any comments.**  If you are familiar with the pthread device, I suggest starting [here](https://github.com/new2f7/pocl/compare/f01c81866dc23e5508557b573ec232d6eaf63212...new2f7:tbb_pthread).

I verified that with the tbb device produces no new errors while running the `conformance_micro` and `conformance_mini` test suits (tests 166 `conformance_compiler` and 412 `conformance_relationals_shuffle` from `conformance_mini` fail with the pthread device as well).
<!--- You can see the changes of my final TBB device compared to my initial renamed copy of the pthread device -->

---

### Notes/Questions before merging this PR:
1) To implement the tbb device, I created a copy of the pthread device and made the necessary changes in a minimal intrusive manner. This led to some files being almost identical except some minor adjustments like renaming (especially in tbb.cc vs. pthread.c and tbb_utils.h/.cc vs. pthread_utils.h/.c). For example, when https://github.com/franz/pocl/commit/de42b3a05ce1090c32a7c53f34432acbb8affa8d got merged into the master I applied it to the tbb device as well.
2) <s>I included a third party CMake module (MIT licence) to locate the TBB library (Apache 2.0 licence). Depending on the [progress at TBB](https://github.com/oneapi-src/oneTBB/issues/6), this module might be removed in the future.</s>
3) I've seen different copyright claim styles: "[pocl developers](https://github.com/pocl/pocl/blob/master/lib/CL/devices/pthread/pthread_scheduler.c#L3)", [a person](https://github.com/pocl/pocl/blob/master/lib/CL/devices/pthread/pthread.c#L4), [an institution](https://github.com/pocl/pocl/blob/master/lib/CL/devices/pthread/pocl-pthread.h#L3), [or both](https://github.com/pocl/pocl/blob/master/lib/CL/devices/pthread/pocl-pthread_scheduler.h#L4). What is the preferred way of claiming copyright at the beginning of each file? For me personally, I'm fine with "PoCL developers". Should I update the year when making changes to a file?

---

### Background:

I'm a master computer science student at Humboldt University Berlin and implemented the tbb device as part of my study project supervised by @noma at Zuse Institute Berlin. <s>There exists a paper at an internal draft stage.</s> **Update 2021-05-12**: [paper](https://dl.acm.org/doi/10.1145/3456669.3456698) and [presentation](https://www.youtube.com/watch?v=hauQGDW6n-8) are now available.


![hexciton](https://user-images.githubusercontent.com/14283108/102473983-632b1680-4058-11eb-92c3-5736d6f01e63.png)
The kernel with the highest performance from each kernel group of the [hexciton_benchmark](https://github.com/noma/hexciton_benchmark) (25 iterations excluding one warmup iteration). The error bars indicate minimum and maximum values observed.

<!-- img src="https://user-images.githubusercontent.com/14283108/100175572-f5635300-2ece-11eb-8bd3-0d39a146dfa2.png" width="60%" alt="hexciton_benchmark chart" -->

![hexciton](https://user-images.githubusercontent.com/14283108/100175580-f85e4380-2ece-11eb-86a4-09588a46c080.png)
Selected operations from the [op_benchmark](https://github.com/noma/op_benchmark) (20480 data_elements,10000 iterations, 45 kernel_runs, 5 kernel_warmups). The error bars indicate the lowest and the highest average runtime observed across all vectorisation modes.

<!-- img src="https://user-images.githubusercontent.com/14283108/100175580-f85e4380-2ece-11eb-86a4-09588a46c080.png" width="60%" alt="op_benchmark chart" -->

pjaaskel commented:
Very nice! Thanks for your willingness to contribute back to the open source. I'm looking forward for finding the time to review this.

Oblomov commented:
This is possibly a stupid question, but would it be possible to actually have the TBB device be something like a subclass of the pthread device, in order to minimize code duplication?

pjaaskel commented:
@Oblomov I agree. This is what is done between basic and several other drivers (at least pthread), the ugly C style of "inheritance".

pjaaskel commented on doc/sphinx/source/tbb.rst:
```diff
@@ -0,0 +1,42 @@
+~~~~~~~~~~~~~~~~
+  Use ``POCL_DEVICES=tbb`` to select the TBB device.
+
+  Optionally, set the ``POCL_TBB_PARTITIONER`` environment variable to one of
```
Is there always at most one TBB device which controls all CPUs? PoCL allows defining a list of devices via POCL_DEVICES and parameters per device id. What happens if there are more than one TBB devices? Error out?

new2f7 replied:
It is strongly recommended to use at most on TBB device. I added this to the [documentation](https://github.com/new2f7/pocl/commit/6efa2d06a76e0d9982fb693269069c011b56b55f#diff-fb6a8d163fc0c1e73f9d455ba2b8eccf979def2873d43e04d4a547f9d76f090aR38).

pjaaskel commented on lib/CL/devices/tbb/tbb.cc:
```diff
@@ -0,0 +1,392 @@
+};
+
+void
+pocl_tbb_init_device_ops(struct pocl_device_ops *ops)
```
Can you do the "C-style inheritance" here and just assign pthread/basic functions to those that are identical or can be "overriden" by first calling it and then modifying after?

new2f7 replied:
See 3c81f27c58d9abfc58af9be860413dc8a0fd6ff1 and subsequent commits.

pjaaskel commented on lib/CL/devices/tbb/tbb.h:
```diff
@@ -0,0 +1,46 @@
+/* OpenCL device using the Intel TBB library (derived from the pthread device).
+
+   Copyright (c) 2011 Universidad Rey Juan Carlos
```
Check the copyright.

new2f7 replied:
Addressed in 2933f1eac17bad78390f97cbdd47e413191706c2.

pjaaskel reviewed (changes requested):
Thanks again for the PR.  The main problem is the abundance of copy-pasted code. Could we invent a nice way to share code with pthread to avoid this? E.g. add the TBB-modified parts as macro protected regions and then define-enable them in the tbb parts or just call the pthread functions instead of copying them. Now this organization leads to more maintenance burden due to duplicated code.

new2f7 commented:
Thanks for the comments so far. I'm aware that the duplicated code segments are difficult to maintain. I'm willing to work on this issue but it might take some time until I'll implement the proposed changes.

pjaaskel commented:
One option would be to rename 'pthread' to 'multithread'/'mt'/'multicore'/mc' and just implement the TBB as an alternative code path for its thread scheduler in it. If the TBB works, I don't know if there is an use case to use both the pthread and the TBB drivers simultaneously. Pthread solely is useful for multithreaded systems without TBB.

new2f7 commented:
Sorry for the long delay with this PR, I was busy finalising the associated [paper ](https://dl.acm.org/doi/10.1145/3456669.3456698)and [presentation](https://www.youtube.com/watch?v=hauQGDW6n-8), which are now available.

In addition to the requested changes above, I've added the following changes:
- 02c2b813ec250fdc9286d7d8504a50123c9b1590 Use a generic pocl_build_hash function for basic, pthread, tbb devices
- 13a1b7b4355b3cafbd0902a69a5eb6b937141e4b Merge pthread_utils.h/.c and tbb_utils.h/.cc into new files called common_utils.h/.c
- I replaced the 3rd party [FindTBB.cmake](https://github.com/justusc/FindTBB) (which [does not support TBB 2021](https://github.com/justusc/FindTBB/issues/16)) with the binary integration according to TBB's [official documentation](https://github.com/oneapi-src/oneTBB/tree/master/cmake#tbbconfig---integration-of-binary-packages).

I also prepared [another commit](https://github.com/new2f7/pocl/commit/36469d72b64bd519b82fd8bf643116638a5295e6) to reduce duplicate code. **To finish it, I would like to know:**
- Why is [the basic device using](https://github.com/pocl/pocl/commit/aca5aff5d157ed2a60e330c984ef3d373de93fef) the official Khronos vendor ID but [the pthread device isn't](https://github.com/pocl/pocl/blob/master/lib/CL/devices/pthread/pthread.c#L196)?
- Maybe consider using the Khronos vendor ID [as fallback in cpuinfo.c](https://github.com/pocl/pocl/blob/d906f0124a7b0b3191d86e4390fd55fb42442db0/lib/CL/devices/cpuinfo.c#L69)? This seems to be the cleanest solution to me.
- How to [handle multiple instances of the same device](https://github.com/pocl/pocl/blob/5f806ca03228f6c4ab4a3a3042dd4475913125a5/lib/CL/devices/pthread/pthread.c#L201) in this case? Is it a problem if multiple instances use the same ID?

If you have any other comments to the code or the things mentioned above, please let me know.

new2f7 commented on CMakeLists.txt:
```diff
@@ -1095,11 +1102,17 @@ if(ENABLE_ACCEL_DEVICE)
 endif()
 
+# The tbb device is implemented in C++ and requires the Intel TBB library.
+if(ENABLE_TBB_DEVICE)
```
Done in 1aac987943e60b92bfcd68d63ae89f305944102a.

pjaaskel commented:
Looks much better now! We'll take a closer look in the group and test it out.

* Congrats on the IWOCL paper! Can you add a link to it in the http://portablecl.org/publications.html ? The web pages are in the repo. 

* Please add yourself to the copyright lines in the significantly edited files.

* If TBB is enabled, I believe it should be used by default instead of pthread. Does your patch do that?

> Why is the basic device using the official Khronos vendor ID but the pthread device isn't?

No reason to my knowledge. The fallback sounds reasonable. Is the vendor id per (physical) device or just device type (the vendor of it)?. I think there's an extension called device uuid or such that is for identifying the physical device.

jansol commented:
As per [this footnote in the spec](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_API.html#_footnotedef_6) is specific to the *OpenCL implementation*, so PoCL should use the official Khronos vendor ID. For the CPU device implementations that seems fairly straightforward. I guess this is mainly meant for informational purposes and for easy filtering of known problematic implementations on the application level (for example IIRC the Cycles renderer in Blender explicitly checks that OpenCL devices have the AMD vendor ID and essentially forces NVIDIA devices to use CUDA instead).

For actually identifying a device there is indeed [an extension](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_Ext.html#cl_khr_device_uuid).

pjaaskel commented on lib/CL/devices/tbb/tbb_scheduler.cc:
```diff
@@ -0,0 +1,435 @@
+  kernel_run_command *my_k;
+public:
+  void operator()( const tbb::blocked_range3d<size_t>& r ) const {
+    kernel_run_command *k = my_k;
```
Please describe briefly in a function comment what the method does and when it's called.

new2f7 replied:
Done in 8acf01b.

pjaaskel commented on lib/CL/devices/tbb/tbb_scheduler.cc:
```diff
@@ -0,0 +1,435 @@
+    for (size_t x=r.pages().begin(); x!=r.pages().end(); x++) {
+      for (size_t y=r.rows().begin(); y!=r.rows().end(); y++) {
+        for (size_t z=r.cols().begin(); z!=r.cols().end(); z++) {
+          pocl_set_default_rm();
```
does it have to set it before each call or is per thread enough?

new2f7 replied:
According to the already existing comment, before/after each call.

pjaaskel commented on lib/CL/devices/tbb/tbb_scheduler.cc:
```diff
@@ -0,0 +1,435 @@
+    }
+
+    if (position > 0) {
+      write (STDOUT_FILENO, pc.printf_buffer, position);
```
Should it dump to STDOUT or use the printf buffer mechanism of OpenCL 1.2 and dump only when debug is on, like with pthread?

new2f7 replied:
This line and if-statement are an [exact copy](https://github.com/pocl/pocl/blob/master/lib/CL/devices/pthread/pthread_scheduler.c#L316) from the pthread device, which AFAICT also dumps to `STDOUT`?

franz replied:
Yep this is a copy of the pthread code, which uses a per-thread printf buffer, and dumps the output to STDOUT after a kernel is finished running.

pjaaskel replied:
OK, seems I misremembered the default implementation.

pjaaskel commented on lib/CL/devices/tbb/tbb_scheduler.cc:
```diff
@@ -0,0 +1,435 @@
+static void
+tbb_exec_command (kernel_run_command *run_cmd)
+{
+  /* Note: Grain size variation could be allowed for each dimension individually. */
```
This if..else seems redundant. Can you just set the grain_size parameter to the default of 1 if it's not set by the user and avoid the else?

new2f7 replied:
I understand that this nested if-else statements is very clumsy, but what if TBB decides to change its default behaviour of `grain_size:=1`? However, if you have a strong opinion on this, I'll change it.

pjaaskel replied:
Good point,  if it's not well defined.

pjaaskel commented on lib/CL/devices/tbb/tbb_scheduler.h:
```diff
@@ -0,0 +1,46 @@
+/* OpenCL device using the Intel TBB library (derived from the pthread device).
+
+   Copyright (c) 2015 Ville Korhonen, Tampere University of Technology
```
check the (c)

pjaaskel reviewed (changes requested):
Device interop is important feature/benefit of PoCL. Can you add a smoke test to the test suite which does ping-pong between at least two different device drivers, e.g. tbb+basic (if tbb is enabled)? There should be some examples like that already, IIRC?

Did you run the tier1 tests with tbb as default already? Does the conformance pass?

new2f7 commented:
Thank you very much for your remarks. I look forward to implementing them.

> Did you run the tier1 tests with tbb as default already? Does the conformance pass?

I'm currently running `conformance_suite_full`. Suits `micro` and `mini` pass. I can run the `tier1` tests afterwards, but I'm using a weird environment for the longer running stuff and it took me ages to figure out [how](https://github.com/new2f7/pocl/commit/352517aa0eebdc666c34047e45e1896a66cf1033) to [run](https://github.com/franz/OpenCL-CTS/compare/cl12_trunk...new2f7:patch-3?expand=1) `conformance_suite_full`.

I'm not very familiar with the copyright stuff. Maybe you could answer question 3 from my initial post or point to some literature?

> 3) I've seen different copyright claim styles: "[pocl developers](https://github.com/pocl/pocl/blob/master/lib/CL/devices/pthread/pthread_scheduler.c#L3)", [a person](https://github.com/pocl/pocl/blob/master/lib/CL/devices/pthread/pthread.c#L4), [an institution](https://github.com/pocl/pocl/blob/master/lib/CL/devices/pthread/pocl-pthread.h#L3), [or both](https://github.com/pocl/pocl/blob/master/lib/CL/devices/pthread/pocl-pthread_scheduler.h#L4). What is the preferred way of claiming copyright at the beginning of each file? For me personally, I'm fine with "PoCL developers".

pjaaskel commented:
> I've seen different copyright claim styles: "pocl developers", a person, an institution, or both. What is the preferred way of claiming copyright at the beginning of each file? For me personally, I'm fine with "PoCL developers".

That should be the one that really owns the copyrights of the contributed work. If you are paid by an organization and the copyrights are transferred to them, they own the copyrights and you should ensure that you can license that code with MIT. In Universities it usually boils down to the project consortium agreements etc. where it's said which licenses must be used etc. for the code written within the project's funds etc. If it's clearly from one owner only then it's preferred to put that one (the person and the organization), PoCL developers have been used when there have been multiple smaller contributors, or the author has been shy, but the convention is not written in stone.

Kazhuu commented:
Hello! I was doing some testing with TBB device and [polybench](https://github.com/cavazos-lab/PolyBench-ACC) OpenCL kernels. I put up a graph below about the execution times with Intel OpenCL and PoCL with both pthread and TBB devices. The execution time is normalized to Intel OpenCL execution time. Meaning that Intel execution time is always 1 and lower is better.

![image](https://user-images.githubusercontent.com/11016113/118168865-fbbf9a00-b430-11eb-8ecb-0d095ecb4834.png)

As can be seen from the graph some kernels execute a bit faster and some a bit slower. What caught my attention is `gramschmidt` kernel increased execution time when using TBB device. I'm not 100% sure what causes this but in the past execution time for this kernel used to take around 44 seconds before I [fixed](https://github.com/pocl/pocl/commit/d6a5395a3f600c5893feed16da53def8d29cb2b1) the synchronization to all previous events in case of in-order queue which was not required. Meaning that when in-order queue was used, pocl did synchronization to all previous events with locking, which slowed down the total execution time a lot. With above fix the execution time for the kernel dropped to ~0.08 seconds. I think the issue is that the `gramschmidt` kernel spams the OpenCL api quite heavily to enqueue kernels. Don't is this related this at all but just pointing out.

noma commented:
@Kazhuu thanks for the benchmark results.  Have you tried playing around with explicitly setting `POCL_TBB_PARTITIONER` as described in the [doc](https://github.com/new2f7/pocl/blob/6efa2d06a76e0d9982fb693269069c011b56b55f/doc/sphinx/source/tbb.rst)?

Kazhuu commented:
I used command `POCL_DEVICES=tbb POCL_TBB_PARTITIONER=auto ./gramschmidt.exe` to use TBB device and set the partitioner.

In polybench you can set the data set size in the header file. Above result I was using `MINI_DATASET` with `gramschmidt` kernel. I retested with `SMALL_DATASET` data set size and actually now TBB executes faster than pthread. Below table of lowest execution times I observed in seconds.

| pthread        | TBB+auto           | TBB+affinity  | TBB+simple | TBB+static |
| ------------- |-------------| -----| ---- |---- |
| 14.501403 | 14.001868 | 12.684936 | 13.981414 | 12.654978 |

So it seems that TBB is better in this case when the data set size gets big enough.

new2f7 commented on lib/CL/devices/tbb/tbb_scheduler.cc:
```diff
@@ -0,0 +1,435 @@
+  /* alloc local memory for all threads making sure the memory for each thread is aligned. */
+  scheduler.printf_buf_size = (1 + (scheduler.printf_buf_size - 1) / MAX_EXTENDED_ALIGNMENT) * MAX_EXTENDED_ALIGNMENT;
+  scheduler.local_mem_size = (1 + (scheduler.local_mem_size - 1) / MAX_EXTENDED_ALIGNMENT) * MAX_EXTENDED_ALIGNMENT;
+  scheduler.printf_buf_global_ptr = reinterpret_cast<uchar*> (pocl_aligned_malloc (MAX_EXTENDED_ALIGNMENT, scheduler.printf_buf_size * scheduler.num_tbb_threads));
```
I applied the clang formatter script to `tbb_scheduler.cc`: c591b07

new2f7 commented on doc/sphinx/source/tbb.rst:
```diff
@@ -0,0 +1,49 @@
+
+Use the ``POCL_DEVICES`` environment variable to select the TBB device, i. e.
+``POCL_DEVICES=tbb``. It is strongly recommended to create only one TBB device
+as the TBB device has no subdevice support and always uses all cores.
```
I added a commit which sets the tbb device as default device and prints a warning if a user requests to use more than one tbb device (0c99ec0). However, I would prefer to not stop adventurous users, who perhaps knows better than I what TBB is doing internally, from creating more than one tbb device.

new2f7 commented on lib/CL/devices/common_utils.h:
```diff
@@ -30,10 +30,9 @@
-#pragma GCC visibility push(hidden)
-#endif
-
+/* NOTE: Some entries in this struct are only required by the pthread device and
```
Implemented this and subsequent comments in b3d1707.

new2f7 commented on lib/CL/devices/tbb/tbb_scheduler.cc:
```diff
@@ -0,0 +1,435 @@
+      POCL_MSG_PRINT_GENERAL ("TBB: using a grain size of %u\n", scheduler.grain_size);
+    }
+
+  scheduler.selected_partitioner = NONE;
```
This would bake TBB's current default policy into PoCL. I would argue against this, because TBB's default policy [has changed in the past](https://github.com/oneapi-src/oneTBB/blob/v2020.3/CHANGES#L2192). Note that TBB's auto_partitioner is an actual partitioner and not a substitute for automatically choosing another partitioner.

new2f7 commented on lib/CL/devices/tbb/tbb_scheduler.cc:
```diff
@@ -0,0 +1,435 @@
+
+/* Internal functions */
+
+class WorkGroupScheduler {
```
See c591b07 (same as above). The mess is up to me, because this file is a copy of a C file with a sole inserted C++ class, compiled with a C++ compiler (TBB is a C++ library).

new2f7 commented on doc/sphinx/source/tbb.rst:
```diff
@@ -0,0 +1,49 @@
+==========
+
+The TBB device uses the Intel Threading Building Blocks open source (Apache 2.0)
+library for scheduling and presents an alternative to the pthread device.
```
Implemented this and subsequent remarks with 6c81528.

new2f7 commented:
I addressed all previous remarks with the above commits or added a comment why I haven't implemented them (yet).

## question about subdevice support

How important is subdevice support for PoCL? Are there common use cases out there?

To achieve rudimentary support, one could read the maximum number of threads to use from an environment variable and pass it to [task_arena.initialize()](https://github.com/new2f7/pocl/blob/tbb_pthread/lib/CL/devices/tbb/tbb.cc#L102) (similar to [POCL_MAX_PTHREAD_COUNT](https://github.com/new2f7/pocl/blob/master/lib/CL/devices/pthread/pthread.c#L185)).

## conformance tests results

|suite|#tests that fail with `pthread`|#tests that _additionally_ fail with `tbb`|
|-----|-----------------------------|--------------------------------------|
|micro|0|0|
|mini|2|0|
|full|11|0|

The following tests from `conformance_suite_mini` failed:
```
166 - conformance_compiler (Failed)
412 - conformance_relationals_shuffle (Failed)
```

The following tests from `conformance_suite_full` failed:
```
171 - conformance_compiler (Failed)
360 - conformance_math_fmod (Failed)
362 - conformance_math_frexp (Failed)
392 - conformance_math_pow (Failed)
393 - conformance_math_pown (Failed)
394 - conformance_math_powr (Failed)
395 - conformance_math_remainder (Failed)
396 - conformance_math_remquo (Failed)
398 - conformance_math_rootn (Failed)
417 - conformance_relationals_shuffle (Failed)
439 - conformance_vecalign (Failed)
```

Logs for `conformance_suite_full` (I reran the failed tests with `ctest --rerun-failed` as suggested by `ctest`):
[conformance_full_pthread.log](https://github.com/pocl/pocl/files/6492463/conformance_full_pthread.log)
[conformance_full_pthread_rerun_failed.log](https://github.com/pocl/pocl/files/6492465/conformance_full_pthread_rerun_failed.log)
[conformance_full_tbb.log](https://github.com/pocl/pocl/files/6492466/conformance_full_tbb.log)
[conformance_full_tbb_rerun_failed.log](https://github.com/pocl/pocl/files/6492467/conformance_full_tbb_rerun_failed.log)

## performance

@Kazhuu Thanks for trying the `tbb` device. I haven't evaluated the performance of the `tbb` device in scenarios other than the ones described in the paper. However, I can describe theoretical scenarios, in which the `pthread` device should be faster than the `tbb` device:

As far as I understand the `pthread` device, the execution of commands and kernels can overlap, because each thread constantly pulls for [commands](https://github.com/new2f7/pocl/blob/master/lib/CL/devices/pthread/pthread_scheduler.c#L466) and [kernels](https://github.com/new2f7/pocl/blob/master/lib/CL/devices/pthread/pthread_scheduler.c#L447) when work is done. For the `tbb` device, there is just one so called _meta thread_ that does this and in case of kernels, executes them in parallel by calling the TBB library. Thus. this behaviour means the `tbb` device cannot execute commands in parallel and the execution of kernels will not overlap.

This bottleneck could be widened by creating more _meta threads_ which effectively execute commands and kernels in parallel. However, this would result in a (much) more complex implementation that also requires an answer to the question how TBB behaves when calling `parallel_for()` multiple times while other ones are still running.

I haven't worked towards such a solution, because the scenarios I was looking at benefited from the current solution and if a workload corresponds to the cases described above, the `pthread` device can still be used.

I hope these remarks help assessing the `tbb` device.

new2f7 commented:
> Device interop is important feature/benefit of PoCL. Can you add a smoke test to the test suite which does ping-pong between at least two different device drivers, e.g. tbb+basic (if tbb is enabled)? There should be some examples like that already, IIRC?

There is a conformance test which utilises two devices (`conformance_events_ooo`). It passes when there is one `pthread` and one `tbb` device (default before 0c99ec0). It does not fail if there is just one device. Could you point me to the mentioned examples if you wish further testing?

franz commented:
> How important is subdevice support for PoCL? Are there common use cases out there?

It becomes important when using PoCL on with machines with multiple NUMA nodes (CPU sockets). In that case you need affinity + subdevices to use the machine efficiently. Granted it's not a very common usecase, but IIRC there were a few people using it, we had a few related bugreports. But IMO it's not necessary to deal with subdevices in the first PR; it's fine have no subdevice support in the TBB device driver for now.

I've looked at the code and it looks reasonable, i only have some minor nitpicks:

1) In `check_cmd_queue_for_device()` - the for loop (DL_FOREACH) is unnecessary (that code was there for subdevices).

2) the TBB shipped in ubuntu doesn't have a CMake file, but it has a pkg-config file; TBB could be searched using `pkg_check_modules` as fallback after `find_package`. Something like this should work:

```
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -407,7 +407,18 @@ endif()
 
 if(ENABLE_TBB_DEVICE)
   # related documentation: https://github.com/oneapi-src/oneTBB/tree/tbb_2020/cmake#binary-package-integration
-  find_package(TBB REQUIRED tbb)
+  find_package(TBB)
+  if(NOT TBB AND PKG_CONFIG_FOUND)
+    pkg_check_modules(TBB tbb>=2015)
+    if(TBB_FOUND)
+      set(TBB_IMPORTED_TARGETS "${TBB_LIBRARIES}" CACHE STRING "TBB libraries" FORCE)
+    endif()
+  endif()
+
+  if(NOT TBB_IMPORTED_TARGETS)
+    message(FATAL_ERROR "Can't find TBB libraries")
+  endif()
+
   # The tbb device depends on the pthread device which depends on the basic device. ENABLE_HOST_CPU_DEVICES enables both.
   set(ENABLE_HOST_CPU_DEVICES 1)
 endif()
```

Aside these, LGTM. Thanks for working on this.

pjaaskel commented:
Otherwise this now looks good and I started testing it (with Michal's fixes on top).

I noticed this breaks build when there are no host devices enabled. I found it in my local CUDA-only build. This is because moving the build hash and default extension stuff to common.c and common_utils.c and assuming they are always there. Also I noticed the build system doesn't treat tbb as a "host device".

I mean:
```
option(ENABLE_HOST_CPU_DEVICES "Add host CPUs as OpenCL devices (basic and pthread)." ON)
```
tbb should be considered a host CPU device, so this setting is now misleading.  The build host device hash for example should be introduced for all CPU device driver types (basic, pthread, tbb).

pjaaskel commented:
Hi - just wanted to check, what is the status/plan of this PR? It would be great to get this in sooner than later to get it some testing.

new2f7 commented:
Hey, I'm sorry, I'm currently not able to work on this PR. Some time ago, I have started some minor work locally which addresses the above comments from late May (Michal's fixes). I'll try to get this work into this PR this weekend but I cannot promise anything. Apart from these changes, I consider this PR ready for merge.

I had a plan to evaluate the performance of the TBB device on an ARM system but this plan is delayed at least until mid August.

pjaaskel commented:
Any updates on this? Would be nice to get this cleaned and pulled in.

isuruf commented:
I rebased this PR on top of master branch. There are 2 tests failing though.

pjaaskel commented:
Thanks @isuruf  for working on this useful feature. Did you check our comments in this thread and try to address them?

new2f7 commented:
@isuruf Thank you for looking into my stale work. Which two tests of which test suite are failing?

isuruf commented:
Two tests in pocl's internal test-suite run by `ctest`.
```
The following tests FAILED:
        113 - runtime/clCreateSubDevices (Failed)
        157 - EinsteinToolkit_SubDev (Subprocess aborted)
Errors while running CTest
```

I guess the failures are because of no subdevice suppport?

isuruf commented:
Any suggestions on how to fix those tests?

@pjaaskel, I looked through the suggestions, but I don't understand what to do in

> In check_cmd_queue_for_device() - the for loop (DL_FOREACH) is unnecessary (that code was there for subdevices).

franz commented:
@isuruf

Yes the tests fail b/c of the subdevice support, we need an extra check in them and exit with a "skip" exit code (IIRC 77).

WRT the `check_cmd_queue_for_device`: it has a for loop that doesn't loop, it just returns the first value. So the whole body can be replaced with `retval = scheduler.work_queue; if (retval) DL_DELETE(scheduler.work_queue, retval); return retval` or even inlined.

new2f7 commented:
I looked into my notes from 9 months ago and found some unfinished code. I'll just paste it here in case anyone finds it useful. Thanks again for looking into my unfinished work!

```
# Unfinished Patch (not working)

diff --git a/lib/CL/devices/tbb/tbb_scheduler.cc b/lib/CL/devices/tbb/tbb_scheduler.cc
index 0ddbc060..c2d6c7a1 100644
--- a/lib/CL/devices/tbb/tbb_scheduler.cc
+++ b/lib/CL/devices/tbb/tbb_scheduler.cc
@@ -273,17 +273,6 @@ static kernel_run_command *pocl_tbb_prepare_kernel(void *data,
   return (run_cmd);
 }

-/* Note: Using a double linked list is probably not necessary any more. */
-static _cl_command_node *check_cmd_queue_for_device() {
-  _cl_command_node *cmd;
-  DL_FOREACH(scheduler.work_queue, cmd) {
-    DL_DELETE(scheduler.work_queue, cmd)
-    return cmd; // return first cmd
-  }
-
-  return NULL;
-}
-
 static void tbb_exec_command(kernel_run_command *run_cmd) {
   /* Note: Grain size variation could be allowed for each dimension
    * individually. */
@@ -366,9 +355,10 @@ static int tbb_scheduler_get_work() {
 RETRY:
   do_exit = scheduler.meta_thread_shutdown_requested;

-  /* execute a command if available */
-  cmd = check_cmd_queue_for_device();
-  if (cmd) {
+  /* check if command is available */
+  if (scheduler.work_queue) {
+    /* command available, remove from queue and execute */
+    DL_DELETE(scheduler.work_queue, cmd)
     POCL_FAST_UNLOCK(scheduler.wq_lock_fast);

     assert(pocl_command_is_ready(cmd->event));
@@ -382,12 +372,13 @@ RETRY:
     }

     POCL_FAST_LOCK(scheduler.wq_lock_fast);
-  }

-  /* if no command was available, sleep */
-  if ((cmd == NULL) && (do_exit == 0)) {
-    pthread_cond_wait(&scheduler.wake_meta_thread, &scheduler.wq_lock_fast);
-    goto RETRY;
+  } else {
+    /* no command available, sleep */
+    if (do_exit == 0) {
+      pthread_cond_wait(&scheduler.wake_meta_thread, &scheduler.wq_lock_fast);
+      goto RETRY;
+    }
   }

   POCL_FAST_UNLOCK(scheduler.wq_lock_fast);


# Draft comment for Github PR 

> 1. In `check_cmd_queue_for_device()` - the for loop (DL_FOREACH) is unnecessary (that code was there for subdevices).

I'm aware of this issue. The current solution is a workaround because utlist.h does not provide something like DL_REMOVE_FIRST.

> 2. the TBB shipped in ubuntu doesn't have a CMake file, but it has a pkg-config file; TBB could be searched using `pkg_check_modules` as fallback after `find_package`. Something like this should work:

I could not reproduce the problem with Linux Mint 20/20.1, which is based on Ubuntu 20.04. I installed `libtbb2` and `libtbb-dev`. However, it works with and without your patch thus I added it to this PR.
```

jansol commented:
> ...TBB shipped in ubuntu doesn't have a CMake file...

Looks like TBB has been shipping the necessary cmake files since version [21.1.1](https://github.com/oneapi-src/oneTBB/releases/tag/v2021.1.1). With Ubuntu 22.04 LTS right around the corner I'm not sure if there is much point in going out of our way to detect older versions as [2021.5.0 appears to already be packaged there](https://packages.ubuntu.com/search?keywords=tbb) under the names `libtbb12` and `libtbb-dev`.

On older systems `find_package` already allows specifying a custom path for a given library so it should be possible to build a more recent version and simply point CMake at it with `-DTBB_DIR=/path/to/tbb`?

pjaaskel commented:
I believe all relevant work from this branch have been now refined and pulled in. Please reopen if not so.
